source/slang/slang-check-stmt.cpp

```diff
@@ -239,17 +239,57 @@ namespace Slang
         subContext.checkStmt(stmt->body);
     }
 
+    SeqStmt* SemanticsStmtVisitor::validateCaseStmts(SwitchStmt* stmt, DiagnosticSink* sink)
+    {
+        auto blockStmt = as<BlockStmt>(stmt->body);
+        if (!blockStmt)
+            return nullptr;
+
+        auto seqStmt = as<SeqStmt>(blockStmt->body);
+        if (!seqStmt)
+            return nullptr;
+
+        bool hasDefaultStmt = false;
+        HashSet<Val*> caseStmtVals;
+        for (auto& sStmt : seqStmt->stmts)
+        {
+            // check that all case tags are unique
+            if (auto caseStmt = as<CaseStmt>(sStmt))
+            {
+                if (auto caseVal = tryConstantFoldExpr(caseStmt->expr, ConstantFoldingKind::CompileTime, nullptr))
```

Comment:
When we visit case stmt, we can fold the constant int to a Val and store that Val on the case stmt. then when we are here, we can simply grab the folded Val from the checked caseStmt.

Currently our compiler cannot fold a enum value as a IntVal, but we will plan to support that soon. When that is supported, we should also be diagnosing an error if a case value cannot be folded into Val.

Reply:
could you please elaborate on the "store that Val on the case stmt" ?
Case stmt has a member expr, which has a value, but it is private.

Reply:
<img width="513" alt="image" src="https://github.com/shader-slang/slang/assets/2652293/a0f55685-84f8-47c2-9189-c126597e419c">

I mean add a `Val* exprVal` field in `CaseStmt` class.
Then in `SemanticsStmtVisitor::visitCaseStmt`, try fold `stmt->expr` into `Val` and store it in `stmt->exprVal`.

Note the TODO in `SemanticsStmtVisitor::visitCaseStmt`:
```
        // TODO(tfoley): Need to coerce to type being switch on,
        // and ensure that value is a compile-time constant
```
Here we will have the opportunity to issue an error when folding into IntVal failed.

If you do so then in `SemanticsStmtVisitor::visitSwitchStmt`, the case stmt would have been checked after `subContext.checkStmt(stmt->body);`, and you should be able to grab the caseStmt's `exprVal` and add it to the hash set.

Reply:
thanks Yong. Was unsure about adding a new member  to the CaseStmt.